suite/tests/linux/clone.c

```diff
@@ -185,15 +229,127 @@ run_with_exit(void)
     exit(run(NULL));
 }
 
-void *p_tid, *c_tid;
+/* Wrapper to invoke clone that supports passing stack_ptr == NULL.
+ * While glibc has a C wrapper for clone(), it prohibits passing NULL for the
+ * stack pointer.
+ * For architectures that we don't yet support, this calls clone() instead;
+ * this way create_thread() and create_thread_clone3() are symmetrical.
+ * Currently, this supports a fcn that does not return and calls exit() on
+ * its own.
+ */
+int
+make_clone_syscall(ptr_uint_t flags, void *stack, pid_t *parent_tid, pid_t *child_tid,
+                   void *tls, void (*fcn)(void))
+{
+    /* Everything is done in assembler to safely manage the potential
+     * stack switch or non-switch. */
+    int result;
+#if defined(X86) && defined(X64)
+    /* Note: Syscall preserves all registers except rax, rcx and r11. */
+    /* Syscall args:
+     * rax = syscall#
+     * rdi = flags
+     * rsi = stack
+     * rdx = parent_tid
+     * r10 = child_tid
+     * r8 = tls */
+    asm("movq %[fcn], %%r15\n\t"
+        "movq %[flags], %%rdi\n\t"
+        "movq %[stack], %%rsi\n\t"
+        "movq %[parent_tid], %%rdx\n\t"
+        "movq %[child_tid], %%r10\n\t"
+        "movq %[tls], %%r8\n\t"
+        "mov %[syscall_num], %%eax\n\t"
+        "syscall\n\t"
+        "testq %%rax, %%rax\n\t"
+        "jnz clone_parent\n\t"
+        "xorq %%rbp, %%rbp\n\t"
+        "call *%%r15\n\t"
+        "clone_parent:\n\t"
+        "mov %%eax, %[result]\n\t"
+        : [result] "=r"(result)
+        : [syscall_num] "i"(SYS_clone), [flags] "rm"(flags), [stack] "rm"(stack),
+          [parent_tid] "rm"(parent_tid), [child_tid] "rm"(child_tid), [tls] "rm"(tls),
+          [fcn] "rm"(fcn)
```

Comment:
We have dynamorio_syscall() (DR's asm syscall wrapper) in the drlibc library explicitly so that tests can use it; you can see several uses in suite/tests/ for examples.  Could we move dynamorio_clone() into drlibc so it can be used here to avoid needing any of this assembly and giving immediate support for all supported architectures?

Reply:
Cool, I wasn't aware of dynamorio_clone. It will need to be augmented to handle passing NULL for the stack ptr, and we'll need a clone3 variant, for each architecture.

Ergo, immediate support for all supported architectures isn't possible, but it's still a reasonable way to go.